---
title: "Mosaic_Benchmark_Figures"
author: "Ade, Camille, and Nate"
date: "`r Sys.Date()`"
output: 
    bookdown::html_document2:
        toc: true
        toc_float: true
        df_print: paged
        code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(patchwork)
library(viridis)
library(ggpubr)
library(ggupset)
```

## Load and Tidy Input Data

### Variant VAFs DF for Figure 2
```{r}
## Loading manually curated variants data
chrom_order <- paste0("chr", c(1:22, "X", "Y"))
curated_variants <- read_tsv(
    here("data", "igv_curated_variants_RM.tsv")
) %>%
    dplyr::arrange(AF) %>%
    mutate(
        index = row_number(),
        CHROM = factor(CHROM, levels = chrom_order)
    ) %>%
    dplyr::arrange(CHROM, POS, REF, ALT)

benchmark_vars <- curated_variants %>%
    filter(RM_KEEP_REMOVE == "KEEP") %>%
    select(CHROM, POS, REF, ALT)

curated_variants_longer <- pivot_longer(curated_variants,
    cols = c(
        "BGI_mut_freq",
        "Element_mut_freq",
        "Pacbio_mut_freq",
        "Illumina_mut_freq"
    ),
    names_to = "Method",
    values_to = "VAF"
) %>%
    mutate(Method = str_replace(Method, "_mut_freq", "")) %>%
    mutate(Method = gsub("Illumina", "Illumina 300X", Method)) %>%
    mutate(Method = gsub("BGI", "BGI 100X", Method)) %>%
    mutate(Method = gsub("Element", "Element 136X", Method)) %>%
    mutate(Method = gsub("Pacbio", "Pacbio 108X", Method))

common_variants <- read_tsv(here("data", "RM_VAF_fig_70cand_updated.tsv")) %>%
    arrange(AF) %>%
    mutate(index = row_number())

curated_variants_longer_pacbio_sep <- read_tsv(here("data", "curated_variants_longer_pacbio_sep.tsv"))

benchmark_source_comp_df <- read_tsv(
    here("data", "ortho_support", "combined_bamread_count.tsv")
) %>%
    right_join(benchmark_vars)
```


### In Silico LOD Data
```{r}
insilico_lod_df <- list.files(here("data/insilico_lod"), full.names = TRUE) %>%
    set_names(., str_extract(., "(?<=af)[:digit:]*(?=_result)")) %>%
    map_dfr(read_csv, .id = "AF") %>%
    mutate(AF = as.numeric(AF))
```

## Figures

### Figure 2
```{r}
ortho_vafs_df <- curated_variants_longer %>%
    mutate(Variant_Category = case_when(
        RM_KEEP_REMOVE == "KEEP" ~ "**Included** in Benchmark",
        RM_KEEP_REMOVE != "KEEP" ~ "**Excluded** from Benchmark",
        TRUE ~ "ERROR"
    ))

(curated_ortho_vafs_fig <- ggplot(
    ortho_vafs_df,
    aes(x = index, y = VAF)
) +
    annotate("rect",
        xmin = -Inf, xmax = Inf,
        ymin = 0.05, ymax = 0.3,
        alpha = 0.4, fill = "gray"
    ) +
    geom_line(aes(group = index), linewidth = 0.5) +
    geom_point(aes(fill = Method), shape = 21, color = "grey40") +
    scale_color_viridis() +
    guides(fill = guide_legend(
        nrow = 1,
        position = "bottom",
        direction = "horizontal",
        title = "Sequencing Method"
    )) +
    labs(x = "Strelka2 VAF", y = "VAF") +
    theme_bw() +
    theme(
        strip.text = ggtext::element_markdown(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()
    ) +
    facet_wrap(
        ~Variant_Category,
        ncol = 1
    )
)
```

```{r}
ggsave(
    filename = here("figures", "fig2_curated-variants.pdf"),
    plot = curated_ortho_vafs_fig,
    height = 6, width = 8
)
```

### Figure 3

```{r currated_VAF, message = FALSE}
(curated_variants_hist <- curated_variants %>%
    mutate(combined_VAF = x_ci_no_Dup / n_ci_no_Dup) %>%
    ggplot(aes(
        x = combined_VAF
    )) +
    geom_vline(
        xintercept = c(0.05, 0.3),
        linetype = "dashed", color = "black"
    ) +
    geom_histogram(
        color = "black",
        position = "identity", bins = 40, na.rm = TRUE
    ) +
    labs(
        x = "Variant Allele Frequency (VAF)",
        y = "# of Variants"
    ) +
    theme_bw() +
    theme(
        strip.text = ggtext::element_markdown()
    ) +
    facet_wrap(~RM_CURATION,
        ncol = 1,
        labeller = labeller(RM_CURATION = c(
            "MOSAIC" = "**Included** in Benchmark",
            "NOT MOSAIC" = "**Excluded** from Benchmark"
        ))
    )
)
```


```{r}
ggsave(
    plot = curated_variants_hist,
    filename = here("figures", "fig3_curated-variants-VAF.pdf"),
    height = 6, width = 4
)
```

### Figure 4

```{r}
(source_vaf_comp <- benchmark_source_comp_df %>%
    filter(
        str_detect(platform, "element|pacbio"),
        !str_detect(platform, "long"),
        !str_detect(platform, "sequel")
    ) %>%
    mutate(tech = str_remove(platform, "_.*")) %>%
    group_by(dna_source, tech, CHROM, POS, REF, ALT) %>%
    summarise(VAF = sum(alt_count) / sum(depth)) %>%
    mutate(
        idx = paste0(CHROM, POS, REF, ALT),
        dna_source = factor(dna_source, levels = c("nonRM", "RM")),
        tech = factor(tech, levels = c("element", "pacbio"), labels = c("Element", "PacBio Revio"))
    ) %>%
    ggpaired(
        x = "dna_source", fill = "tech",
        y = "VAF",
        id = "idx",
        facet.by = "tech",
        line.color = "gray",
        line.size = 0.4,
        palette = c("#7CAE00", "#C77CFF"),
    ) +
    stat_compare_means(paired = TRUE) +
    ylim(0, 0.4) +
    theme(legend.position = "none")
)
```

```{r}
ggsave(
    plot = source_vaf_comp,
    filename = here("figures", "fig4_VAF-RM-v-nonRM.pdf")
)
```

### Figure S1

```{r}
(lod_fig <- insilico_lod_df %>%
    ggplot() +
    geom_vline(aes(xintercept = 10), linetype = 1) +
    geom_vline(aes(xintercept = 5), linetype = 2) +
    geom_path(aes(x = AF, y = METRIC.Recall, color = Filter, linetype = Type)) +
    geom_point(aes(x = AF, y = METRIC.Recall, fill = Filter, shape = Type)) +
    scale_shape_manual(values = c(21, 22)) +
    theme_bw() +
    labs(x = "Allele Frequency (%)", y = "Recall") +
    theme(legend.position = "bottom"))
```

```{r}
ggsave(
    plot = lod_fig,
    filename = here("figures", "figS1_insilico-lod.pdf")
)
```

### Figure S4
```{r}
(curated_vafs_hist <- ggplot(curated_variants_longer, aes(x = VAF, fill = Method)) +
    geom_histogram(color = "black", alpha = 0.7) +
    facet_wrap(Method ~ ., scales = "fixed", ncol = 1) +
    scale_fill_manual("Method",
        breaks = c("BGI 100X", "Element 136X", "Illumina 300X", "Pacbio 108X"),
        values = c("#F9746D", "#7CAE00", "#00BEC4", "#C77CFF")
    ) +
    theme_bw() +
    theme(legend.position = "none") +
    xlab("Variant Allele Frequency (VAF)") +
    ylab("Number of Manually Curated Variants"))
```

```{r}
ggsave(
    plot = curated_vafs_hist,
    filename = here("figures", "figS4_curated-VAFs.pdf"),
    height = 8, width = 6
)
```

### Figure S5
```{r}
(bench_vafs_hist <- curated_variants_longer %>%
    filter(RM_KEEP_REMOVE == "KEEP") %>%
    ggplot(aes(x = VAF, fill = Method)) +
    geom_histogram(color = "black", alpha = 0.7) +
    facet_wrap(Method ~ ., scales = "fixed", ncol = 1) +
    scale_fill_manual("Method",
        breaks = c("BGI 100X", "Element 136X", "Illumina 300X", "Pacbio 108X"),
        values = c("#F9746D", "#7CAE00", "#00BEC4", "#C77CFF")
    ) +
    theme_bw() +
    theme(legend.position = "none") +
    xlab("Variant Allele Frequency (VAF)") +
    ylab("Number of Mosaic Variants"))
```

```{r}
ggsave(
    plot = bench_vafs_hist,
    filename = here("figures", "figS5_benchmark-VAFs.pdf"),
    height = 8, width = 6
)
```

# Session Information
## System Information

```{r}
sessioninfo::platform_info()
```

## Package Versions
```{r}
sessioninfo::package_info() %>%
    filter(attached = TRUE) %>%
    select(package, loadedversion, date, source) %>%
    knitr::kable(booktabs = TRUE, row.names = FALSE)
```
